app/scripts/modules/core/src/presentation/modal/showModal.tsx

```diff
@@ -0,0 +1,107 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+import { UIRouterContextComponent } from '@uirouter/react-hybrid';
+
+import { Modal } from './Modal';
+
+/** The Modal content Component will be passed these two props */
+export interface IModalComponentProps<C = any, D = any> {
+  // Close modal with a result value (i.e., OK button)
+  closeModal?(result?: C): void;
+
+  // Dismiss/reject modal (i.e., Cancel button)
+  dismissModal?(result?: D): void;
+}
+
+interface ModalCloseResult<T> {
+  status: 'CLOSED';
+  closeResult: T;
+}
+
+interface ModalDismissResult<T> {
+  status: 'DISMISSED';
+  dismissResult?: T;
+}
+
+export type IModalResult<C, D> = ModalCloseResult<C> | ModalDismissResult<D>;
+
+/**
+ * An imperative API for showing a react component as a modal.
+ *
+ * example:
+ * const MyComponent = ({ closeModal, dismissModal }) => {
+ *   <h1>Modal Contents!</h1>
+ *   <button onClick={() => closeModal('A')}>Choice A</button>
+ *   <button onClick={() => closeModal('B')}>Choice B</button>
+ *   <button onClick={() => dismissModal('cancelled')}>Cancel</button>
+ * }
+ *
+ * ...
+ *
+ * showModal(MyComponent).then(result => {
+ *   this.setState({ result });
+ * });
+ *
+ * @param ModalComponent the component to be rendered inside a modal
+ * @param componentProps to pass to the ModalComponent
+ * @returns {Promise<IModalResult<C, D>}
+ */
+export const showModal = <P, C = any, D = any>(
+  ModalComponent: React.ComponentType<P & IModalComponentProps<C, D>>,
+  componentProps?: P,
+): Promise<IModalResult<C, D>> =>
+  new Promise<IModalResult<C, D>>(resolve => {
+    let mountNode = document.createElement('div');
+    let show = false;
+
+    function onAfterClose() {
+      if (!mountNode) {
+        return;
+      }
+
+      ReactDOM.unmountComponentAtNode(mountNode);
+      mountNode = null;
+    }
+
+    const handleResultWith = (resultHandler: (result: C | D) => IModalResult<C, D>) => (result: C | D) => {
+      if (!mountNode) {
+        return;
+      }
+
+      resolve(resultHandler(result));
+      // Switch `show` to false to trigger exit transition and call onExitComplete
+      show = false;
+      render();
+    };
+
+    const handleClose = handleResultWith((result: C) => ({ status: 'CLOSED', closeResult: result }));
+    const handleDismiss = handleResultWith((result: D) => ({ status: 'DISMISSED', dismissResult: result }));
+    const handleRequestClose = () => handleDismiss(null);
+
+    function render() {
+      ReactDOM.render(
```

Comment:
Have you considered using [react portal](https://reactjs.org/docs/portals.html#usage) here instead?

Reply:
Good Q, the `Modal` component itself [actually renders in a portal](https://github.com/spinnaker/deck/blob/master/app/scripts/modules/core/src/presentation/modal/Modal.tsx#L33-L45) — the reason this imperative API requires a real mount point is because it's intentionally _not_ rendering the `Modal` component into an existing component tree. Portals are great, but they inherently have to be mounted into an existing component tree.

Reply:
Ah, right. You had mentioned earlier that you didn't explicitly want these kinds of modals to be in the same react tree.